XiEditor/FileSystemItem.swift

```diff
@@ -0,0 +1,111 @@
+// Copyright 2018 Google Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Cocoa
+
+final class FileSystemItem {
+
+    fileprivate var relativePath: String!
+    fileprivate var parent: FileSystemItem?
+    lazy fileprivate var children: [FileSystemItem] = self.getChildren()
+
+    static let rootFileSystemItem = URL(fileURLWithPath: "/")
+
+    var fullPath: String {
+        guard
+            let parent = self.parent,
+            let url = URL(string: parent.fullPath)?.appendingPathComponent(self.relativePath)
+            else { return self.relativePath }
+
+        return url.absoluteString
+    }
+
+    var url: URL {
+        return URL(string: self.fullPath)!
+    }
+
+    var fileURL: URL {
+        return URL(fileURLWithPath: self.fullPath)
+    }
+
+    var fileType: String {
+        return self.fileURL.pathExtension
+    }
+
+    var isDirectory: Bool {
+        var isDir: ObjCBool = false
+        FileManager.default.fileExists(atPath: self.fullPath, isDirectory: &isDir)
+        return isDir.boolValue
+    }
+
+    convenience init() {
+        self.init(path: FileSystemItem.rootFileSystemItem.path)
+    }
+
+    init(path: String, parent: FileSystemItem? = nil) {
+        self.relativePath = URL(fileURLWithPath: path).lastPathComponent
+        self.parent = parent
+    }
+
+    // MARK: - Public methods
+
+    public func getChild(at index: Int) -> FileSystemItem {
+        return children[index]
+    }
+
+    public func getNumberOfChildren() -> Int {
+        return children.count
+    }
+
+    public func getName() -> String {
```

Comment:
Alternatively, for this `getName()` and `getNumberOfChildren` methods, I'd just use a computed property here.

For example, `getName()` would be something like `var name: String { return self.relativePath }`. This is more "Swift", in my opinion.